### PR TITLE
Fix Taproot input size estimation

### PR DESCRIPTION
**Pull request by benthecarman**
Should be step 1 of https://github.com/lightningnetwork/lnd/issues/6873

### CODE REVIEW

File: wallet/txsizes/size.go [hunk 1]
```diff
@@ -83,6 +83,37 @@ const (
...
+	//   - OP_DATA_32
+	//   - 32 bytes pubkey
+	P2TRPkScriptSize = 1 + 1 + 32
+
+	// P2TROutputSize is the serialize size of a transaction output with a
+	// P2TR output script. It is calculated as:
+	//
+	//   - 8 bytes output value
+	//   - 1 byte compact int encoding value 22
+	//   - 22 bytes P2PKH output script
```
Comment (guggero) [inline wallet/txsizes/size.go:99]: Comment should say 34 bytes P2TR output script.

File: wallet/txsizes/size.go [hunk 2]
```diff
@@ -125,6 +156,14 @@ const (
 	//   - 1 wu compact int encoding value 33
 	//   - 33 wu serialized compressed pubkey
 	RedeemP2WPKHInputWitnessWeight = 1 + 1 + 73 + 1 + 33
+
+	// RedeemP2TRInputWitnessWeight is the worst case weight of
+	// a witness for spending P2TR outputs. It
+	// is calculated as:
+	//
+	//   - 1 wu compact int encoding value 65
+	//   - 64 wu DER signature + 1 wu sighash
```
Comment (guggero) [inline wallet/txsizes/size.go:165]: nit: Schnorr or BIP340 Signature, not DER anymore.

File: wallet/txsizes/size.go [hunk 3]
```diff
@@ -156,11 +195,11 @@ func EstimateSerializeSize(inputCount int, txOuts []*wire.TxOut, addChangeOutput
 }
 
 // EstimateVirtualSize returns a worst case virtual size estimate for a
-// signed transaction that spends the given number of P2PKH, P2WPKH and
+// signed transaction that spends the given number of P2PKH, P2TR, P2WPKH and
 // (nested) P2SH-P2WPKH outputs, and contains each transaction output
 // from txOuts. The estimate is incremented for an additional P2PKH
 // change output if addChangeOutput is true.
-func EstimateVirtualSize(numP2PKHIns, numP2WPKHIns, numNestedP2WPKHIns int,
+func EstimateVirtualSize(numP2PKHIns, numTRIns, numP2WPKHIns, numNestedP2WPKHIns int,
```
Comment (guggero) [inline wallet/txsizes/size.go:202]: nit: call this `numP2TRIns` just to follow the same pattern as the others?

**Review by guggero — CHANGES_REQUESTED** [review]
Thanks for the fix!
The code itself looks very good :100: 

Unfortunately go modules makes this PR a bit hard. Or to be more precise, the changes need to be done in two PRs, because `wallet/txsizes` is its own Go module that needs to be tagged separately:
1. Create a PR that only contains the changes in `wallet/txsizes`, get that merged and we push a new tag `wallet/txsizes/v1.2.0`.
2. Create a PR that uses the new `wallet/txsizes/v1.2.0` version in `wallet/txauthor/go.mod` and `author.go`.
3. Get the second PR merged and push a new tag `wallet/txauthor/v1.3.0`.

**Comment by benthecarman** [comment]
Okay updated to only fix the sizes

**Review by guggero — APPROVED** [review]
LGTM, thanks a lot :tada:

**Comment by guggero** [comment]
@Roasbeef gentle reminder to push a new tag `wallet/txsizes/v1.2.0` after merging this, so we can get the second part in.

**Comment by benthecarman** [comment]
pushed a typo fix

File: wallet/txsizes/size.go [hunk 4]
```diff
@@ -125,6 +156,14 @@ const (
...
 	//   - 33 wu serialized compressed pubkey
 	RedeemP2WPKHInputWitnessWeight = 1 + 1 + 73 + 1 + 33
+
+	// RedeemP2TRInputWitnessWeight is the worst case weight of
+	// a witness for spending P2TR outputs. It
+	// is calculated as:
+	//
+	//   - 1 wu compact int encoding value 65
+	//   - 64 wu BIP-340 schnorr signature + 1 wu sighash
+	RedeemP2TRInputWitnessWeight = 1 + 65
```
Comment (Roasbeef) [inline wallet/txsizes/size.go:166]: FWIW, this'll be _slightly_ off in the case that sighash default is used (implicit alias for sighash all).

File: wallet/txsizes/size.go [hunk 5]
```diff
@@ -156,11 +195,11 @@ func EstimateSerializeSize(inputCount int, txOuts []*wire.TxOut, addChangeOutput
 }
 
 // EstimateVirtualSize returns a worst case virtual size estimate for a
-// signed transaction that spends the given number of P2PKH, P2WPKH and
+// signed transaction that spends the given number of P2PKH, P2TR, P2WPKH and
 // (nested) P2SH-P2WPKH outputs, and contains each transaction output
 // from txOuts. The estimate is incremented for an additional P2PKH
 // change output if addChangeOutput is true.
-func EstimateVirtualSize(numP2PKHIns, numP2WPKHIns, numNestedP2WPKHIns int,
+func EstimateVirtualSize(numP2PKHIns, numP2TRIns, numP2WPKHIns, numNestedP2WPKHIns int,
```
Comment (Roasbeef) [inline wallet/txsizes/size.go:202]: TBH, we could prob make lnd's weight estimator into a module and use it here, then we wouldn't need to add a new arg here each time we have a diff output type (ofc at this pace it doesn't matter much). 

Just a note, don't think it's something we need to do rn.

**Review by Roasbeef — APPROVED** [review]
LGTM 🦜

**Comment by Roasbeef** [comment]
@benthecarman had to manually approve the CI run, failing on the linter with this:
```
Error: rpcserver.go:11:2: SA1019: package io/ioutil is deprecated: As of Go 1.16, the same functionality is now provided by package io or package os, and those implementations should be preferred in new code. See the specific function documentation for details. (staticcheck)
	"io/ioutil"
	^
Error: cmd/sweepaccount/main.go:9:2: SA1019: package io/ioutil is deprecated: As of Go 1.16, the same functionality is now provided by package io or package os, and those implementations should be preferred in new code. See the specific function documentation for details. (staticcheck)
	"io/ioutil"
```

You can just use `os` now for that.

**Comment by Roasbeef** [comment]
Actually I think it's a case of the linter+default version updating and flagging some older code.

File: wallet/txsizes/size.go (hunk 4 above)
Comment (benthecarman) [reply wallet/txsizes/size.go:166]: Yeah, all of them said the worst case scenario so I included the sighash byte

**Comment by benthecarman** [comment]
@Roasbeef pushed a fix for the lint errors, they seem unrelated to this PR however

**Comment by Roasbeef** [comment]
@benthecarman yeh they're unrelated, but the linter wants to flag them for some reason....

Seeing this still:
```
Error: rpc/legacyrpc/server.go:14:2: SA1019: package io/ioutil is deprecated: As of Go 1.16, the same functionality is now provided by package io or package os, and those implementations should be preferred in new code. See the specific function documentation for details. (staticcheck)
```

And this now for the test failure: 
```
./rpcserver.go:11: imported and not used: "io/ioutil"
```

**Comment by benthecarman** [comment]
Okay I believe I removed all instances of `io/ioutil` now

**Comment by Roasbeef** [comment]
Cool, once this is merged in, I'll now longer have to manually approve each run

**Comment by Roasbeef** [comment]
Almost there....
```
Error: chain/bitcoind_events_test.go:5:2: SA1019: package io/ioutil is deprecated: As of Go 1.16, the same functionality is now provided by package io or package os, and those implementations should be preferred in new code. See the specific function documentation for details. (staticcheck)
```

**Comment by benthecarman** [comment]
Ah needed to rebase, should be good to go now

**Comment by Roasbeef** [comment]
Ok we're g2g now!

**Comment by Roasbeef** [comment]
Ok the new tag has been pushed: https://github.com/btcsuite/btcwallet/releases/tag/wallet%2Ftxsizes%2Fv1.2.0

We can do the second part now: 
> Create a PR that uses the new wallet/txsizes/v1.2.0 version in wallet/txauthor/go.mod and author.go.

**Comment by Roasbeef** [comment]
Slight mix up, you need to use: `wallet/txsizes/v1.2.1`

**Comment by benthecarman** [comment]
https://github.com/btcsuite/btcwallet/pull/811

**Pull request merged**